Add Beta Upload Activity - User Interface Redesign

davelopez opened the pull request:
This PR introduces a **new, beta upload experience** for Galaxy, available alongside the existing upload dialog. The new interface moves from a modal-based approach to a dedicated activity panel with a method-first workflow and clearer separation between different import mechanisms.

The interface is clearly marked with a **Beta** to indicate it's in the testing phase. The goal is to gather user feedback on the new upload flow before replacing the traditional Upload Dialog in a future release.

> [!NOTE]
> **To provide feedback** comment on discussion issue: https://github.com/galaxyproject/galaxy/discussions/21340

<img width="431" height="948" alt="image" src="https://github.com/user-attachments/assets/170a74f8-aab1-424f-a8f7-49a8b70c24c9" />


## How to Activate the Beta Upload

The beta upload interface is **opt-in** and can be accessed through the activity bar:

1. Click the **"More" (⋯)** button at the bottom of the activity bar (left sidebar)
2. Find **"Beta Upload"** in the additional activities list
3. Click to enable or "add to favorites"
4. Drag the activity icon to the top section for easy access and close the "More" menu when done
5. Click the **"Beta Upload"** icon in the activity bar to open the new upload panel

<img width="432" height="319" alt="image" src="https://github.com/user-attachments/assets/8623d799-0b78-4430-8a22-0f1c3ef15b8a" />


## Key Improvements

### 1. **Method-First Discovery**

**Before**: Single modal with 4 tabs (Regular/Composite/Collection/Rules Builder) requiring users to understand Galaxy's upload taxonomy upfront.

**Now**: A set of **explicitly separated import methods** presented as a searchable list with clear names, descriptions, and contextual tips:

- **Upload from Computer** - Local file selection with drag & drop
- **Paste/Fetch Data** - Direct content pasting or URL imports
- **Choose Remote Files** - Cloud storage and FTP browsers
- **Data Libraries** - Import from shared data libraries
- **Browse ZIP Archive** - Extract and import from ZIP files
- **Data Source Tools** - Launch Galaxy data source tools (e.g., NCBI, UCSC, ENA)
- **Import Entire History** - Clone complete analysis histories
- **Import Workflow** - Load workflow definitions
- **Rule-based Import** - Advanced bulk import (requires Advanced Mode toggle)

Each method includes **contextual tips** intended to guide users toward the most appropriate import option and best practices.

<img width="1191" height="877" alt="image" src="https://github.com/user-attachments/assets/97977eeb-0ee9-4593-8bcf-2d0abfddd749" />

### 2. **Staging Area for Local and Remote Files**

**Before**: The elements selected in the upload modal were lost if navigation occurred.

**Now**:

- **Staging across method interactions** - Upload elements selected via upload methods can be reviewed and adjusted before starting the upload and persist across navigation within the upload panel (please note: local files cannot persist across full browser reloads due to browser security restrictions)
- **Incremental batch building** - Add more files before starting an upload
- **Per-file configuration** - Adjust name, datatype, genome build, and conversion options prior to upload
- **Clear visual feedback** - Table-based review of staged files with inline validation

<img width="859" height="870" alt="image" src="https://github.com/user-attachments/assets/24e737bf-8b2e-4e2a-bddb-9d87c2f4f0cc" />


### 3. **Integrated Collection Creation**

- **Pre-upload collection configuration** - Declare intent before uploading for simple collection types
- **Automatic pairing detection** - Common paired-end naming conventions are detected where applicable
- **Real-time validation** - Prevents invalid configurations (e.g., odd file count for pairs)
- **Automatic creation** - Collection built immediately after successful uploads

<img width="856" height="971" alt="image" src="https://github.com/user-attachments/assets/94ed821d-eb51-4e0a-ac58-9312b8404628" />


### 4. **Upload Progress Tracking**

- **In-panel progress indicator** - Shows active uploads in the panel header
- **Dedicated progress page** - A full progress view is available at `/upload/progress`
- **Batch grouping** - Collection uploads are displayed together with aggregate progress
- **Status updates** for individual uploads and collection-related operations
- **Rich file details**:
    - Upload progress percentage
    - File size and type indicators
    - Target history with quick-switch links
    - Elapsed time since upload started
    - Clear error messages with actionable guidance when possible/available

- **Persistent history** - Completed uploads remain visible until manually cleared

<img width="1218" height="687" alt="image" src="https://github.com/user-attachments/assets/77f7f767-253a-456e-a2e8-da66c274604f" />


### 5. **Advanced Mode Toggle**

- **Progressive disclosure** - Advanced features hidden by default
- **Per-user preference** - Setting persisted across sessions (same browser)
- **Filtered methods** - Rule-based Import only shown in Advanced Mode
- **Simplified controls** - POSIX line ending options hidden unless needed

### 6. **Target History Selection**

**Before**: Uploads always went to current history.

**Now**:

- **History selection** - Choose destination history before uploading
- **Visual confirmation** - Target history clearly displayed during upload
- **Quick switching** - Links to switch to target history from progress view
- **Per-upload selection** - Target history is selected as part of the upload flow

<img width="1152" height="651" alt="image" src="https://github.com/user-attachments/assets/1da86e13-6bd7-41e5-b6f2-7b1d220e1604" />


## Technical Notes

### Architecture

- **Modular component system** - Upload methods independently registered and loaded
- **TypeScript throughout** - Type-safe implementation with comprehensive interfaces
- **Composable business logic** - Reusable upload queue, staging, and collection creation composables
- **Client-side routing** - Deep-linkable upload workflows (`/upload`, `/upload/local-file`, `/upload/progress`)

## Known Limitations

⚠️ **The following features are not yet implemented in the beta upload interface**:

1. **Composite datatypes** - Complex multi-file datatypes are not currently handled. Use the legacy upload dialog for these cases.

2. **Some specialized upload scenarios** - If you encounter issues with specific workflows, please report them in the [discussion issue here](https://github.com/galaxyproject/galaxy/discussions/21340) and use the legacy dialog as a fallback. You are welcome to test the beta upload for general use cases or use the discussion to request specific missing features or provide feedback.

The traditional Upload Dialog remains available and fully functional during the beta period.

## We Need Your Feedback!

This beta release is specifically designed to **gather user feedback** before replacing the legacy upload dialog. We're particularly interested in:

- **Ease of discovery** - Can you find the upload method you need?
- **Workflow efficiency** - Does the new interface save you time?
- **Clarity** - Are the tips and guidance helpful?
- **Error handling** - If uploads fail, are error messages actionable or clear?
- **Collection creation** - Does the integrated workflow make sense?
- **Missing features** - What capabilities from the old dialog do you miss?
- **Pain points** - What's confusing or frustrating?



## How to test the changes?
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [x] Instructions for manual testing are as follows:
  - See the [How to Activate the Beta Upload](#how-to-activate-the-beta-upload) section above

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

davelopez committed:
Add experimental beta upload panel to activity bar

davelopez committed:
Add options to import histories and workflows

davelopez committed:
Splits paste upload into content and links modes

This way we can have more control  about the content, how to validate it, suggest using deferred data, etc.

davelopez committed:
Remove category from upload methods

Name and description already cover the same words for searching so no need for it for now.

davelopez committed:
Adds modal dialogs for guided, advanced, and method-specific uploads

davelopez committed:
Adds fullscreen mode option to GModal component

davelopez committed:
Makes all upload dialogs full screen sized

davelopez committed:
Moves upload panel component to dedicated folder

davelopez committed:
Refactors upload panel to modular upload method system

Introduces a registry-driven upload method architecture, enabling dynamic selection and configuration of upload types. Moves upload logic into modular, async-loaded components. Lays groundwork for easier extensibility, improved user experience, and future integration of new import workflows.

davelopez committed:
Improves upload dialog layout and dynamic component integration

davelopez committed:
Adds prototype local file upload dialog with drag-and-drop UI

davelopez committed:
Adds option to select history during upload

davelopez committed:
Refactors upload handling to simplify event flow

davelopez committed:
Adds reusable tip component with markdown support

Introduces a flexible UI component for displaying informational, warning, success, or danger tips with markdown formatting. Enhances user guidance and maintains consistent styling across the app.

davelopez committed:
Replaces help alerts with GTip component in some upload panels

davelopez committed:
Adds upload state management for new upload panel

davelopez committed:
Adds composable upload service

davelopez committed:
Add upload progress indicator and details modal

Improves user feedback during file uploads by introducing a visual upload progress indicator in the upload panel header and a detailed modal displaying individual file upload statuses, progress, and errors.

davelopez committed:
Adds target history selection for upload methods

davelopez committed:
Adds configurable upload metadata controls to local file uploads

Introduces controls for file extension, dbkey, whitespace, line endings, and deferred upload options to enhance flexibility and consistency for local file uploads. Integrates upload service to handle file enqueueing with user-specified metadata.

davelopez committed:
Implement initial multi-dataset paste upload UI

davelopez committed:
Add basic UI for pasting URLs to upload files with options

davelopez committed:
Remove Collection File Upload from panel

Collections should be offered in each activity when there are more than 1 files selected for upload.

davelopez committed:
Rename variable to improve readability

davelopez committed:
Refactors upload method selection from modal dialog to the central panel

Moves upload method selection to dedicated route.

davelopez committed:
Adds upload method selection index page

davelopez committed:
Refactors upload UI to use component ref

davelopez committed:
Move upload progress to center panel instead of modal dialog

davelopez committed:
Refactor upload progress view navigation and actions

davelopez committed:
Improves upload progress layout readability

Avoiding flickering when the status changes quickly.

davelopez committed:
Uses useFileDrop composable in LocalFileUpload

To ensure the global drop zone does not interfere with the local drop zone.

jmchilton commented on client/src/api/upload.ts:
```diff
@@ -0,0 +1,12 @@
+export type FetchDatasetHash = components["schemas"]["FetchDatasetHash"];
+
+// Union of supported element sources for the basic (non-advanced) panel
+export type UploadElement = FileDataElement | PastedDataElement | UrlDataElement;
```
We already have some type import stuff for this upload stuff in api/tools.ts - I guess I'd like to see this work reuse those imports and maybe the abstractions being setup in there also. Maybe I'm wrong though - I don't feel this way urgently.

davelopez replied:
Yes, definitely! I will reuse the imports from api/tools as much as I can! Thanks for pointing it out!

davelopez committed:
Refactors upload state and service

davelopez committed:
Unifies upload types under api/tools

davelopez committed:
Updates response interface for tool fetch API

davelopez committed:
Refactors upload logic

Modernizes file upload modules with strong typing.
Improves test coverage and updates imports across affected components for consistency.

davelopez committed:
Refactors upload payload logic

Migrates upload payload construction to a strongly-typed module. Replaces legacy, ad-hoc payload building with explicit type-safe builders and helpers, cleaning up API boundaries and reducing conversion ambiguity between UI and API models.

Aims for improved reliability and future extensibility of dataset upload features.

davelopez committed:
Unifies upload utilities and streamlines dataset submission

Refactors upload logic by consolidating payload building, TUS file uploads, and API submission into a single, reusable module. Eliminates legacy code and replaces scattered upload functions with a unified API, improving maintainability and testability.

davelopez committed:
Refactors upload handling to use queue-based composable

Introduces a composable upload queue system that processes uploads sequentially and improves UI state management. Supports clearer progress tracking and validation for different upload modes.

davelopez committed:
Improves filename normalization for URL uploads

Ensures uploaded filenames are cleaned of URL paths, query parameters, and properly decoded, reducing user confusion and preventing malformed names. Adds comprehensive tests for edge cases and malformed input to increase reliability.

davelopez committed:
Updates option keys to use index in select lists

Replaces composite string keys with index-based keys for select options to avoid potential key collisions.

davelopez committed:
Fix default value assignment logic

Replaces nullish coalescing with logical OR to set default values, ensuring fallback assignments for both null and undefined cases.

davelopez committed:
Adds bulk edit for file type and database selection

davelopez committed:
Adds warning icon for extensions with upload warnings

davelopez committed:
Improves file options UI and tooltip clarity

Refines layout and sizing of file options in the upload panel.
Adds informative tooltips to clarify conversion checkboxes.
Enhances user understanding and usability of file upload features.

davelopez committed:
Adds bulk toggle options for file upload settings

davelopez committed:
Improves upload UI with tooltips and clearer labels

davelopez committed:
Improves file upload table UX and validation

davelopez committed:
Resets drop zone state on file drop

Prevents lingering visual feedback after a file is dropped by resetting the drop zone state.

davelopez committed:
Enables sorting by file name and size in upload table

davelopez committed:
Adds automatic collection creation after batch upload

Enables automatic creation of dataset collections (list or list:paired) after successful batch uploads. Introduces logic for pairing files, naming collections, and error handling on collection creation without interrupting individual uploads.

davelopez committed:
Adds UI and logic for creating collections during file upload

Enables users to create named collections (list or list of pairs) from selected files during upload, with input validation, pairing preview, and automatic hiding of source files. Improves usability for managing grouped uploads and reduces manual post-upload steps.

davelopez committed:
Improves memory cleanup for upload batch tracking

Ensures batches are removed from memory when uploads complete or are cleared, preventing orphaned batch references and reducing memory usage.

davelopez committed:
Fix target history ID reactivity

davelopez committed:
Hides history change link for anonymous users

davelopez committed:
Adds help link for dataset collections in upload panel

davelopez committed:
Refactors collection creation into reusable component

davelopez committed:
Makes the "Add More Files" button less prominent

To keep the upload start button the primary action.

davelopez committed:
Adds persistent batch upload tracking and recovery UI

davelopez committed:
Cleans up expanded batch state after clearing uploads

davelopez committed:
Handles missing upload data after page refresh

Prevents collection creation if upload items are missing due to a page refresh, informing users to re-upload or manually create the collection. Improves error handling during batch recovery and clarifies user-facing error messages for unrecoverable states.

davelopez committed:
Improves cleanup of completed upload batches

Ensures batches are removed not only when marked completed but also when all their uploads have been cleared, preventing orphaned or empty batches from remaining in state.

davelopez committed:
Refactors collection creation to use unified state object

davelopez committed:
Centralizes collection type and input type definitions

Moves collection type and input definitions to a shared types module to improve type consistency and reduce duplication across components and composables.

davelopez committed:
Redesigns paste upload UI with table, bulk actions, and validation

davelopez committed:
Auto-scrolls paste table to show newly added items

davelopez committed:
Adds expand/collapse all for pasted content rows

davelopez committed:
Improves table usability with easier content expansion

By making size and content preview cells clickable for expanding content.

davelopez committed:
Redesigns paste links upload with bulk editing and table UI

davelopez committed:
Improves URL validation and user feedback on upload

Enhances URL validation logic to provide more accurate checks and clearer error messages when pasting links for upload. Invalid URLs now prevent uploads, and users receive specific feedback for common mistakes.

davelopez committed:
Refactors upload panels for reusable bulk/batch logic

Introduces composable utilities and shared components for bulk operations and validation across upload panels, reducing duplication and improving maintainability. Moves extension, dbkey, and upload settings header UI into reusable components, centralizes item validation and upload readiness checks, and cleans up code logic. Enhances consistency and simplifies future enhancements to upload features.

davelopez committed:
Refactors upload table cells into reusable components

Replaces inline form controls for name, extension, dbkey, and options in the upload table with dedicated, reusable cell components.

davelopez committed:
Refactors paste content upload table to use shared composables and cells

davelopez committed:
Refactors local file upload to use modular table components

davelopez committed:
Refactors upload methods to centralize item defaults and mapping

Unifies upload item type definitions and default value handling across all upload method components. Introduces composables for managing upload defaults and collection creation state, and utility mappers for converting internal items to upload queue format. Improves maintainability by reducing duplication and clarifying the upload flow from UI to queue.

davelopez committed:
Refactors upload types for better modularity

davelopez committed:
Adds support for remote files upload

Basic implementation using tables.

davelopez committed:
Adds breadcrumb navigation to remote files upload

Improves user navigation in remote file browser by replacing the back button
with a clickable breadcrumb trail, allowing direct access to previous
navigation steps. Enhances clarity and usability, especially when
navigating deep directory structures.

davelopez committed:
Adds reusable breadcrumb navigation component

This can potentially be consolidated with the BreadcrumbHeading component. This version is meant to be used anywhere and not just in the page heading.

davelopez committed:
Refactors breadcrumb navigation to use shared component

davelopez committed:
Prevents outdated remote file loads on rapid navigation

Introduces a sequence-based mechanism to ensure only the latest remote file or directory load request updates the UI. Guards against race conditions and stale data when users navigate quickly, improving reliability and user experience.

davelopez committed:
Prioritizes and marks user-created file sources

Sorts file sources so user-created ones appear first and adds a user icon to visually distinguish them in the UI, improving clarity and discoverability for users managing their own sources.

davelopez committed:
Adds search functionality to remote file upload panel

davelopez committed:
Adds error retry option to remote file upload

Improves error handling with remote files.

davelopez committed:
Fix wrong import

davelopez committed:
Enhances remote file upload metadata display

davelopez committed:
Improves empty state messaging for remote file browser

davelopez committed:
Refines pagination logic and resets on clear

davelopez committed:
Improves remote file details rendering logic

davelopez committed:
Improves breadcrumb accessibility and text handling

davelopez committed:
Add feature support checks for file sources

Introduces utility functions to determine if a file source supports pagination, search, or sorting. Enhances modularity and enables conditional UI or logic based on available features.

davelopez committed:
Adds client-side pagination and search fallback for remote files

Enhances remote file browsing by detecting file source capabilities and providing client-side pagination and search when server-side support is unavailable. Improves user experience and flexibility by ensuring browsing and filtering work consistently across different file sources, with appropriate UI feedback for large loads.

davelopez committed:
Adds select-all checkbox for file selection in remote upload

davelopez committed:
Improves search input handling in upload dialog

Switches search component to use value and input event binding
for better compatibility and reactivity.

davelopez committed:
Improves remote file source connection UX

Moves the "Connect New Remote Source" button to a more visible location when at root, pairing it with clearer guidance text.

davelopez committed:
Adds globe icon for admin-created file sources

Distinguishes admin-created file sources with a globe icon and tooltip, improving clarity between user and global file sources in the UI.

davelopez committed:
Avoids redundant name assignment in upload item creation

And fixes missing provided name for pasted datasets in the old Upload dialog

davelopez committed:
Adds option to hide wizard heading for embedding

Introduces a prop allowing the wizard's default heading to be conditionally hidden.
Enables cleaner embedding of the wizard in contexts where a heading is already provided, improving UI flexibility and preventing redundant headings.

davelopez committed:
Adds API utilities for fetching library and folder data

Introduces functions to retrieve libraries, individual library details,
and contents of folders with support for filtering, sorting, and pagination.
Helps standardize access to library-related backend endpoints and ensures error handling is consistent across calls.

davelopez committed:
Adds data library dataset import from Upload Activity

Enables users to select and import datasets directly from data libraries within the upload activity.

davelopez committed:
Unifies API upload item types and refactors upload logic

davelopez committed:
Refactors upload item lookup with helper function

davelopez committed:
Adds permission icons for datasets in data library upload

Enhances the data library upload panel by displaying lock and shield icons to indicate dataset privacy and restricted access.
Improves clarity for users regarding dataset accessibility and permissions.

davelopez committed:
Adds data source tools option to Upload Activity

davelopez committed:
Adds upload staging for persistent file selection

Implements a staging store to preserve selected upload items for each method, allowing users to recover in-progress uploads across navigation or reloads.

davelopez committed:
Refactors upload staging logic into reusable composable

davelopez committed:
Drop animation in local-files upload

davelopez committed:
Refactors upload method list to use card component

Improves UI consistency.

davelopez committed:
Renames UploadMode type to UploadMethod for clarity

davelopez committed:
Add guided and advanced upload modes with routing

davelopez committed:
Removes Guided and Advanced upload modes temporarily

Removes related components and routes until further development is resumed.

davelopez committed:
Adds advanced mode toggle to upload panel

Introduces a user-specific advanced mode toggle for the upload panel, allowing users to enable or disable advanced upload options. The setting is persisted in local storage and shared across upload components, improving usability and customization.

davelopez committed:
Shows POSIX option only in upload advanced mode

Limits display of the POSIX line endings conversion option to advanced mode in upload panels, improving UI clarity for standard users while keeping advanced features accessible when needed.

davelopez committed:
Adds rule-based import option to upload panel

Introduces a new upload method that allows users to launch the rule-based data import activity directly from the upload panel. Enhances discoverability and streamlines access to rule-based imports.

davelopez committed:
Makes upload methods reactive to advanced mode

Replaces static upload method retrieval with a computed, reactive approach that filters methods based on whether advanced mode is enabled. Enhances upload method type documentation for clarity and future maintainability. Improves UX by only exposing advanced methods when appropriate.

davelopez committed:
Refactors upload method selection to shared component

davelopez committed:
Refactors upload progress UI for better modularity

Extracts file and batch progress display logic into reusable, dedicated utility and component. Simplifies the main upload progress view by removing inline status logic and using a shared UI contract, improving maintainability and consistency of progress indicators.

davelopez committed:
Adds detailed upload and batch info to upload panels

Improves the upload progress UI by displaying per-file and per-batch metadata, such as upload method, deferred status, and target history. Enhances user clarity about upload destinations and statuses, especially when working with multiple upload modes or histories.

davelopez committed:
Improves upload progress view with auto-scroll on new items

Ensures the upload progress panel automatically scrolls to the latest item when new uploads are added, enhancing user visibility.

davelopez committed:
Moves file type icon from badge to row header

davelopez committed:
Adds upload time display to file row UI

davelopez committed:
Unifies upload progress display ordering and structure

Introduces an ordered, unified list combining batch and standalone uploads for consistent progress tracking and rendering. Refactors UI logic to iterate over this new structure, improving maintainability and user clarity by ensuring all uploads are shown in creation order regardless of type.

davelopez committed:
Refactors upload progress UI into modular components

davelopez committed:
Centralizes upload breadcrumb for consistent Beta labeling

davelopez committed:
Hides experimental upload interface from activity list

Temporarily disables visibility of the beta upload feature to prevent general user access while further improvements and testing are underway.

davelopez committed:
Replaces placeholder with history import component

Integrates the actual import functionality to improve usability and move beyond placeholder content in the upload panel.

davelopez committed:
Replaces placeholder with workflow import component

Integrates the actual workflow import functionality by embedding the import component, removing the placeholder and moving towards a functional upload panel.

davelopez committed:
Adds option to hide Start button for custom upload methods

Introduces a configuration flag allowing upload methods to suppress the default Start button when handling their own submission. Improves UI flexibility for methods with embedded forms or wizards.

davelopez committed:
Removes upload size display from progress indicator

davelopez committed:
Fixes progress summary wrapping for better layout

davelopez committed:
Adds navigation for multiple tips in GTip component

davelopez committed:
Adds auto-rotating tips with configurable interval

Introduces automatic rotation for tips, starting with a random tip and allowing the interval to be customized. Enhances user engagement and ensures timely tip visibility without manual interaction. Cleans up rotation timer on unmount to prevent memory leaks.

davelopez committed:
Adds some contextual tips to upload methods UI

davelopez committed:
Updates tip variant prop to use ColorVariant type

davelopez committed:
Adjusts checkbox spacing for advanced mode switch

guerler commented on client/src/utils/upload-queue.js:
```diff
@@ -100,7 +99,7 @@ export class UploadQueue {
                 throw new Error(`Missing target history for upload item [${index}] ${item.fileName}`);
             }
-            const data = uploadPayload([item], item.targetHistoryId);
+            const data = buildLegacyPayload([item], item.targetHistoryId);
```
Will this be dropped once we switch over to the new uploader activities and remove the current upload modal?

davelopez replied:
Yes, that is the intention. This and some other functions and types in the `Legacy Compatibility Layer` section are marked as `deprecated` to be removed once we completely switch over. But before that, we will probably need some kind of "dialog version" of the new upload in case there are places in tool forms or landing requests that require a way to upload without losing the current context in the central panel.
